core/chains/orm.go

```diff
@@ -0,0 +1,246 @@
+	return
+}
+
+func (o *ChainsORM[ID, CFG, CH]) UpdateChain(id ID, enabled bool, config CFG, qopts ...pg.QOpt) (chain CH, err error) {
+	q := o.q.WithOpts(qopts...)
+	sql := fmt.Sprintf(`UPDATE %s_chains SET enabled = $1, cfg = $2, updated_at = now() WHERE id = $3 RETURNING *`, o.prefix)
+	err = q.Get(&chain, sql, enabled, config, id)
+	return
+}
+
+// StoreString saves a string value into the config for the given chain and key
+func (o *ChainsORM[ID, CFG, CH]) StoreString(chainID ID, name, val string) error {
+	s := fmt.Sprintf(`UPDATE %s_chains SET cfg = cfg || jsonb_build_object($1::text, $2::text) WHERE id = $3`, o.prefix)
+	res, err := o.q.Exec(s, name, val, chainID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to store chain config for chain ID %v", chainID)
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected == 0 {
+		return errors.Wrapf(sql.ErrNoRows, "no chain found with ID %v", chainID)
+	}
+	return nil
+}
+
+// Clear deletes a config value for the given chain and key
+func (o *ChainsORM[ID, CFG, CH]) Clear(chainID ID, name string) error {
+	s := fmt.Sprintf(`UPDATE %s_chains SET cfg = cfg - $1 WHERE id = $2`, o.prefix)
+	res, err := o.q.Exec(s, name, chainID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to clear chain config for chain ID %v", chainID)
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected == 0 {
+		return errors.Wrapf(sql.ErrNoRows, "no chain found with ID %v", chainID)
+	}
+	return nil
+}
+
+func (o *ChainsORM[ID, CFG, CH]) DeleteChain(id ID, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+	query := fmt.Sprintf(`DELETE FROM %s_chains WHERE id = $1`, o.prefix)
+	result, err := q.Exec(query, id)
+	if err != nil {
+		return err
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected == 0 {
+		return sql.ErrNoRows
+	}
+	return nil
+}
+
+func (o *ChainsORM[ID, CFG, CH]) Chains(offset, limit int, qopts ...pg.QOpt) (chains []CH, count int, err error) {
+	err = o.q.WithOpts(qopts...).Transaction(func(q pg.Queryer) error {
+		if err = q.Get(&count, fmt.Sprintf("SELECT COUNT(*) FROM %s_chains", o.prefix)); err != nil {
+			return errors.Wrap(err, "failed to fetch chains count")
+		}
+
+		sql := fmt.Sprintf(`SELECT * FROM %s_chains ORDER BY created_at, id LIMIT $1 OFFSET $2;`, o.prefix)
+		err = q.Select(&chains, sql, pg.Limit(limit), offset)
+		return errors.Wrap(err, "failed to fetch chains")
+	}, pg.OptReadOnlyTx())
+
+	return
+}
+
+func (o *ChainsORM[ID, CFG, CH]) EnabledChains(qopts ...pg.QOpt) (chains []CH, err error) {
+	q := o.q.WithOpts(qopts...)
+	chainsSQL := fmt.Sprintf(`SELECT * FROM %s_chains WHERE enabled ORDER BY created_at, id;`, o.prefix)
+	if err = q.Select(&chains, chainsSQL); err != nil {
+		return
+	}
+	return
+}
+
+// NodesORM is a generic ORM for nodes.
+//
+// OPT: shared generic Node = db.Node[whichever.Node]
```

Comment:
Planning to do this next: https://app.shortcut.com/chainlinklabs/story/36673/generic-chain-and-node-types